### PR TITLE
Delete spurious lat/long setting from the audit log

### PR DESCRIPTION
**Pull request by duncanjbrown**
## Context

We have 1.3 million audit entries that consist of the lat/long flipping back and forth:

Screenshot 2021-03-24 at 07 39 59

## Changes proposed in this pull request

- Find all the entries that consist only of lat/long changes
- for each provider, ascertain whether one of these entries originally set the lat/long
- if so, delete all the lat/long-only audit logs for that provider but that one
- if not, delete all the lat/long audit logs for that provider

## Guidance to review

The SQL could definitely do with another pair of eyes! And could we dry-run this somehow?

## Link to Trello card

https://trello.com/c/6vbtx9Nm/427-audit-trial-is-full-of-flip-flops

## Things to check

- [X] This code does not rely on migrations in the same Pull Request
- [X] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [X] API release notes have been updated if necessary
- [X] This code does not rely on the addition/removal of Azure config environment variables in the same Pull Request
- [X] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/docs/environment-variables.md#azure-hosting-devops-pipeline)

### CODE REVIEW

File: app/services/data_migrations/fix_lat_long_flip_flops.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+      ).order(created_at: :asc)
+
+      audit_count = 0
+      total_audits = audits.count
+
+      Provider.find_each do |p|
+        audits_for_provider = audits.where(auditable_id: p.id)
+
+        # if we already had lat/lng when we made the first spurious update, delete every audit
+        if audits.first.audited_changes['latitude'].last.nil?
```
Comment (stevehook) [inline app/services/data_migrations/fix_lat_long_flip_flops.rb:21]:
```suggestion
        if audits_for_provider.first.audited_changes['latitude'].last.nil?
```
Not sure about this, but `audits.first` would always be the first of the audits for all providers I think?

File: app/services/data_migrations/fix_lat_long_flip_flops.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+          audits_for_provider.destroy_all
+
+        # otherwise, delete all the audits except the one that set it for the first time
+        else
+          audits_for_provider.where("audited_changes#>>'{longitude, 1}' is null").destroy_all
+          audits_for_provider.where("audited_changes#>>'{longitude, 0}' is null").offset(1).destroy_all
+        end
```
Comment (stevehook) [inline app/services/data_migrations/fix_lat_long_flip_flops.rb:28]: If the last flip was setting the values to `null` the data would be in a inconsistent state with the audit trail. I guess the data itself would get fixed up the next time the sync happens so it's not a big problem.

Comment (stevehook) [reply app/services/data_migrations/fix_lat_long_flip_flops.rb:28]: I'm not sure the `is null` condition works as intended here. The problem is that `audited_changes#>>'{longitude, 0}'` is `null` for records that don't have the `longitude` key at all.

Maybe something like `audited_changes#>>'{longitude, 1}' is not null AND audited_changes#>>'{longitude, 0}' is null` etc. is needed.

**Review by stevehook — COMMENTED** [review]
I've not been able to test thoroughly but have a couple of inline comments.

File: app/services/data_migrations/fix_lat_long_flip_flops.rb (hunk 2 above)
Comment (duncanjbrown) [reply app/services/data_migrations/fix_lat_long_flip_flops.rb:28]: I _think_ the original query guards against that by checking that the audited_changes includes both latitude and longitude and it also contains no other keys?

**Comment by stevehook** [comment]
For the dry run the only thing I can think of is replacing the actual deletes with log statements for say the first 10 providers. I guess we could then spot check the dry run changes with the actual data. I think it's worth doing.

File: app/services/data_migrations/fix_lat_long_flip_flops.rb (hunk 2 above)
Comment (duncanjbrown) [reply app/services/data_migrations/fix_lat_long_flip_flops.rb:28]: Wouldn't line 26 delete all flips to null?

**Comment by duncanjbrown** [comment]
> For the dry run the only thing I can think of is replacing the actual deletes with log statements for say the first 10 providers. I guess we could then spot check the dry run changes with the actual data. I think it's worth doing.

Good idea — I'll add

File: app/services/data_migrations/fix_lat_long_flip_flops.rb (hunk 1 above)
Comment (duncanjbrown) [reply app/services/data_migrations/fix_lat_long_flip_flops.rb:21]: Good catch!

File: app/services/data_migrations/fix_lat_long_flip_flops.rb (hunk 2 above)
Comment (stevehook) [reply app/services/data_migrations/fix_lat_long_flip_flops.rb:28]: Ah yes I think you're right about the conditions, that's makes sense now!

For the data consistency thing, my only concern was that the lat/lng values would be null immediately after the migration but that's dumb - we fixed the flips to null a while back so all the current data should be set to non-null values (where they ever existed).

**Comment by duncanjbrown** [comment]
Thanks for comments @stevehook — I've updated accordingly.

**Review by stevehook — APPROVED** [review]
:+1: This looks good to me now. The only question I had was whether it's worth limiting the dry run providers to those that have courses that are synched with apply (so that the 50 sample represents the most useful test cases).

**Comment by duncanjbrown** [comment]
Thanks @stevehook. I can't argue with that, will amend before merging!

**Pull request merged**